N°5235 - Add tmp dir is writable check in setup

piRGoif opened the pull request:
If the tmp dir isn't writable then the compilation will crash, with a stack trace and no explicit information.

Example of the setup.log : 
```
2022-09-07 15:13:07 | Info    |       | ##### STEP copy start | SetupLog
2022-09-07 15:13:07 | Info    |       | ##### STEP copy duration: 0s | SetupLog
2022-09-07 15:13:07 | Info    |       | operation 'async_action', peak memory usage. 2729008 | SetupLog
2022-09-07 15:13:07 | Info    |       | ##### STEP compile start | SetupLog
2022-09-07 15:13:07 | Info    |       | Compiling data model. | SetupLog
2022-09-07 15:13:07 | Info    |       | ----> Entering maintenance mode | SetupLog
2022-09-07 15:13:20 | Error   |       | An exception occurred: movedir: the source directory '' is not a valid directory or cannot be read at line 808 in file C:\Dev\wamp64\www\itop-27\setup\setuputils.class.inc.php | SetupLog
2022-09-07 15:13:20 | Ok      |       | Call stack: | SetupLog
2022-09-07 15:13:20 | Ok      |       | #0 C:\Dev\wamp64\www\itop-27\setup\compiler.class.inc.php(186): SetupUtils::movedir(...) | SetupLog
2022-09-07 15:13:20 | Ok      |       | #1 C:\Dev\wamp64\www\itop-27\setup\applicationinstaller.class.inc.php(624): MFCompiler->Compile(...) | SetupLog
2022-09-07 15:13:20 | Ok      |       | #2 C:\Dev\wamp64\www\itop-27\setup\applicationinstaller.class.inc.php(292): ApplicationInstaller::DoCompile(...) | SetupLog
2022-09-07 15:13:20 | Ok      |       | #3 C:\Dev\wamp64\www\itop-27\setup\wizardsteps.class.inc.php(2464): ApplicationInstaller->ExecuteStep(...) | SetupLog
2022-09-07 15:13:20 | Ok      |       | #4 C:\Dev\wamp64\www\itop-27\setup\ajax.dataloader.php(177): WizStepSummary->AsyncAction(...) | SetupLog
2022-09-07 15:13:20 | Info    |       | ##### STEP compile duration: 12.39s | SetupLog
2022-09-07 15:13:20 | Info    |       | operation 'async_action', peak memory usage. 36346328 | SetupLog
```

This PR aims to check on the first step if the temp dir is writable and display a clear message to the iTop admin.

piRGoif committed:
Add tmp dir is writable check in setup
If not present, will crash curing compilation (in \MFCompiler::Compile, when doing `unlink($sTempTargetDir);`)

Molkobain commented:
Accepted in technical review. @piRGoif will test on Windows / Linux env to ensure it does not bring any regression before reviewing it in functional review.

piRGoif commented:
For now, \SetupUtils::CheckWritableDirs only checks under APPROOT :(

piRGoif committed:
Fix didn't detect non writable temp dir
Now tests using a dedicated code, as \SetupUtils::CheckWritableDirs can only check under APPROOT !

piRGoif commented:
Now the readonly is correctly detected !
Rebased the branch by the way O:)

piRGoif commented:
Error message : 

![image](https://user-images.githubusercontent.com/8947448/177366846-588a9e8e-57e5-4b00-bab7-45e47674a66a.png)

And success message : 

![image](https://user-images.githubusercontent.com/8947448/177367052-f761bae7-3838-4e37-8036-e49b981d6b38.png)

Molkobain commented on setup/setuputils.class.inc.php:
```diff
@@ -81,16 +81,23 @@ public static function CheckPhpAndExtensions()
+		if (is_writable($sTmpDir)) {
+			$aResult[] = new CheckResult(CheckResult::INFO, "The temp directory is writable by the application.");
+		} else {
+			$aResult[] = new CheckResult(CheckResult::ERROR, "The temp directory <b>'".$sTmpDir."'</b> is not writable for the application. Change its permission or use another dir (sys_temp_dir option in php.ini).");
```
```suggestion
			$aResult[] = new CheckResult(CheckResult::ERROR, "The temp directory <b>'".$sTmpDir."'</b> is not writable by the application. Change its permission or use another dir (sys_temp_dir option in php.ini).");
```

To be consistent with the success message.

Molkobain commented on setup/setuputils.class.inc.php:
```diff
@@ -81,16 +81,23 @@ public static function CheckPhpAndExtensions()
-		$aResult = array_merge($aResult, $aWritableDirsErrors);
+		$aWritableApprootDirsErrors = self::CheckWritableDirs(array('log', 'env-production', 'env-production-build', 'conf', 'data'));
+		$aResult = array_merge($aResult, $aWritableApprootDirsErrors);
+		// check temp dir (N°5235)
```
```suggestion
		// Check temp dir (N°5235)
```

Molkobain commented:
Nice addition to the preliminary checks ✌

piRGoif committed:
:bulb: Update comment format

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

piRGoif committed:
:speech_balloon: Update tmp writable check error message

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

Molkobain commented:
Discussed during technical review: Why not include the directory test in the already existing `self::CheckWritableDirs()` method? Is it just for the specific error message?

piRGoif commented:
> Discussed during technical review: Why not include the directory test in the already existing `self::CheckWritableDirs()` method? Is it just for the specific error message?

That was what I wrote in my message dated of 5th july : this method is meant only to check dirs under approot for now... I tried first to refactor it, but it made the code complicated so I added the tmp dir check as a standalone test : simplier, much easier to read and understand.

Molkobain commented:
Ok thanks for the clarification, maybe we should add a comment above it to avoid someone to refactor this and put the tmp dir in the above array. What do you think?

piRGoif commented:
Yep, adding a phpdoc block to \SetupUtils::CheckWritableDirs would be a good idea !
I'll do it in the upcoming days

piRGoif commented:
Rebase branch  + force push

piRGoif committed:
:bulb: Add phpdoc for \SetupUtils::CheckWritableDirs
We want to explain this method only checks for dirs under APPROOT

piRGoif commented:
PHPDoc added : 4a026469

piRGoif commented:
Processed today in technical review : good to go for functional review next week !

Hipska commented on setup/setuputils.class.inc.php:
```diff
@@ -1798,21 +1805,31 @@ public static function CheckVersion($sInstalledVersion, $sSourceDir)
+	 *
+	 * @param string[] $aWritableDirs list of dirs to check, relative to APPROOT (for example : `['log','conf','data']`)
+	 *
+	 * @return array full path as key, CheckResult error as value
+	 *
+	 * @uses \CheckResult
```
```suggestion
	 * @return \CheckResult[] full path as key, CheckResult error as value
	 *
```

piRGoif replied:
I'm unsure this is right ? 
For me `\CheckResult[]` would mean there are no keys ? I mean we don't have an associative array ?

Hipska replied:
It says it’s an array, nothing about keys. But that is in the description. 

To me it is more clear than just “array”. I’m only not sure about the correct syntax.

piRGoif replied:
Not sure either :(

The phpdoc reference doesn't help as it does say nothing about associative arrays : https://docs.phpdoc.org/guide/references/phpdoc/types.html#arrays

In the PHPStan syntax I suppose the 3rd syntax is dedicated to associative arrays ? https://phpstan.org/writing-php-code/phpdoc-types#general-arrays
It says more in the Psalm reference : https://psalm.dev/docs/annotating_code/type_syntax/array_types/#generic-arrays

This article does sum up, and list indeed the type[] syntax as a standard indexed array (no keys), and describe both the phpstan and psalm syntax and array shapes : https://dev.to/suckup_de/modern-phpdoc-annotations-11d4

This psalm / phpstan syntax seems to be understood well by PHPStorm, I've pushed ca699a45

Hipska replied:
Yup indeed, if you want to be that specific, then it should become this: `* @return array<string, \CheckResult>`. Otherwise `* @return \CheckResult[]` or `* @return array<\CheckResult>` are sufficient.

piRGoif replied:
I've just updated our [coding standards wiki page](https://www.itophub.io/wiki/page?id=latest:customization:coding_standards#phpdoc)

piRGoif committed:
:bulb: Update \SetupUtils::CheckWritableDirs return type

piRGoif commented:
Rebased & forced pushed !

piRGoif committed:
:bulb: Update \SetupUtils::CheckWritableDirs phpdoc

Molkobain commented:
Functional review: See with @Rorcha07 for possible complications on exotic setups (eg. lodabalancers, netwrk drives, ...)
If any concern raised will be part of 3.1.0, if green light from Stephane then 2.7.8

piRGoif commented:
Discussed with Stephane : we will switch the check to warning so it will become non blocking. 
The information will still be there, and we won't block setup wrongly !

piRGoif commented:
New message when dir isn't writable moved to warning : 

![image](https://user-images.githubusercontent.com/8947448/189954139-eda2e2a5-d50a-40ab-948f-1dc60cfea8c3.png)

And after opening the details : 

![image](https://user-images.githubusercontent.com/8947448/189954190-ccc76a1a-1b20-46f0-9e52-b1c6d776f114.png)

Molkobain commented on setup/setuputils.class.inc.php:
```diff
@@ -81,16 +81,23 @@ public static function CheckPhpAndExtensions()
+		if (is_writable($sTmpDir)) {
+			$aResult[] = new CheckResult(CheckResult::INFO, "The temp directory is writable by the application.");
+		} else {
+			$aResult[] = new CheckResult(CheckResult::ERROR, "The temp directory <b>'".$sTmpDir."'</b> is not writable by the application. Change its permission or use another dir (sys_temp_dir option in php.ini).");
```
Did you push your last modifications for changing the error into a warning? Because it still looks an error here.

Molkobain replied:
Nevermind I just saw it's done bdfe3a3b

piRGoif replied:
Yep, I made a mistake, forgot to push the modification before merging the PR (!!)
To make things a little clearer I've added a reference to this PR in the commit I've pushed after the pr merging.

piRGoif replied:
See in the PR log : 

![image](https://user-images.githubusercontent.com/8947448/190341957-29bc93a3-1f5f-4e23-bc74-63dc85fc3329.png)

Molkobain commented:
_Extra commit under bdfe3a3b to actually push the error to warning change_
